source/common/http/conn_manager_impl.cc

```diff
@@ -202,7 +202,17 @@ void ConnectionManagerImpl::doEndStream(ActiveStream& stream) {
              StreamInfo::ResponseFlag::UpstreamConnectionTermination))) {
       stream.response_encoder_->getStream().resetStream(StreamResetReason::ConnectError);
     } else {
-      stream.response_encoder_->getStream().resetStream(StreamResetReason::LocalReset);
+      // responseDetails are filled when codec encountered parsing error.
+      if (!stream.response_encoder_->getStream().responseDetails().empty()) {
+        if (Runtime::runtimeFeatureEnabled(
+                "envoy.reloadable_features.return_502_for_upstream_protocol_errors")) {
+          stream.response_encoder_->getStream().resetStream(StreamResetReason::ProtocolError);
```

Comment:
Is this correct in all cases? I think not given that response details could come from local reply, etc.? I think we should probably be checking an actual state vs. checking for a non-empty string? I think I had commented on this a while back but it seems even riskier now.

Reply:
Changed this to rely on `streamInfo().hasResponseFlag(StreamInfo::ResponseFlag::UpstreamProtocolError)` flag, which is set by codec client when error is detected.